### ISSUE TITLE
[REG] Basis for classification architecture

### ISSUE THREAD

**Issue by shotnothing** [gugarosa/learnergy#16]
Regarding the basic Bernoulli RBM for classification, could I check what is the basis of your implementation? 
From what I can tell, the training is broken up in two sections:
1. RBM will perform unsupervised learning to minimize energy (RBM::fit), basically like a feature extractor.
2. The whole model is then considered as a 2-layer neural network with a sigmoid layer (RBM::forward) and linear transformation layer.

So, it seems like the RBM is just used as an initializer for the neural network. 

In addition, I noticed that RBM::fit's epoch defaults to 1. Increasing the epoch number does not seem to improve performance. Does this mean that there is no use in training the RBM?

Lastly, was there any prior work that implemented RBM in this way, or is this your original idea? Are there any published results or study into this architecture? Thank you!

**Comment by gugarosa** [comment]
Hello @shotnothing! I hope everything is going well with you.

Exactly. The `applications/bernoulli/rbm_classification.py` is just an example of how to use an RBM as a feature extractor and further classify these features using a Neural Network.

Regarding its performance, there might be a chance that given MNIST is a straightforward task, there is no benefit in extending the pre-training or the additional gains are purely diminishing returns due to the saturated pre-training. There might be a chance that the fine-tuning is messing up with the pre-training learned parameters (high learning rate in the model.parameters() optimizer). Also, there is always the chance that the fine-tuning is the one that is saturated, given the simplicity of the classification head attached on top of the RBM.

@MateusRoder should be able to tell you better where he used this type of architecture. But as far as I know, it is not something "out-of-the-box". It is really a feature extractor with a classification head.

**Comment by gugarosa** [comment]
On the other hand, there are some literatures related to RBMs and discriminative training: http://machinelearning.org/archive/icml2008/papers/601.pdf. And code as well: https://github.com/gugarosa/learnergy/blob/master/learnergy/models/bernoulli/discriminative_rbm.py

This is more "novel" as the discriminative training is taken along the RBM's training, instead of just attaching a Neural Network head.

**Comment by shotnothing** [comment]
@gugarosa Thank you so much for the detailed reply! It corroborates with what I have observed working with this library. I shall take a closer look at the literature you have linked.
